hoodie-client/src/main/java/com/uber/hoodie/index/bloom/HoodieBloomIndex.java

```diff
@@ -0,0 +1,380 @@
+     *
+     * totalSubPartitions : this is deemed safe limit, to be nice with Spark.
+     * inputParallelism : typically number of input file splits
+     *
+     * We pick the max such that, we are always safe, but go higher if say a there are a lot of
+     * input files. (otherwise, we will fallback to number of partitions in input and end up with
+     * slow performance)
+     */
+    private int determineParallelism(int inputParallelism, int totalSubPartitions) {
+        // If bloom index parallelism is set, use it to to check against the input parallelism and take the max
+        int indexParallelism = Math.max(inputParallelism, config.getBloomIndexParallelism());
+        int joinParallelism = Math.max(totalSubPartitions, indexParallelism);
+        logger.info("InputParallelism: ${" + inputParallelism + "}, " +
+                "IndexParallelism: ${" + config.getBloomIndexParallelism() + "}, " +
+                "TotalSubParts: ${" + totalSubPartitions + "}, " +
+                "Join Parallelism set to : " + joinParallelism);
+        return joinParallelism;
+    }
+
+    /**
+     * Load all involved files as <Partition, filename> pair RDD.
+     */
+    @VisibleForTesting
+    List<Tuple2<String, BloomIndexFileInfo>> loadInvolvedFiles(List<String> partitions, final HoodieTable<T> hoodieTable) {
+        // Obtain the latest data files from all the partitions.
+        List<Tuple2<String, HoodieDataFile>> dataFilesList = jsc.parallelize(partitions, Math.max(partitions.size(), 1))
+                .flatMapToPair(partitionPath -> {
+                    java.util.Optional<HoodieInstant> latestCommitTime =
+                            hoodieTable.getCommitTimeline().filterCompletedInstants().lastInstant();
+                    List<Tuple2<String, HoodieDataFile>> filteredFiles = new ArrayList<>();
+                    if (latestCommitTime.isPresent()) {
+                         filteredFiles =
+                                hoodieTable.getROFileSystemView().getLatestDataFilesBeforeOrOn(partitionPath,
+                                        latestCommitTime.get().getTimestamp())
+                                        .map(f -> new Tuple2<>(partitionPath, f))
+                                        .collect(toList());
+                    }
+                    return filteredFiles.iterator();
+                }).collect();
+
+        if (config.getBloomIndexPruneByRanges()) {
+            // also obtain file ranges, if range pruning is enabled
+            return jsc.parallelize(dataFilesList, Math.max(dataFilesList.size(), 1))
+                    .mapToPair(ft -> {
+                        try {
+                            String[] minMaxKeys = ParquetUtils.readMinMaxRecordKeys(ft._2().getFileStatus().getPath());
+                            return new Tuple2<>(ft._1(), new BloomIndexFileInfo(ft._2().getFileName(), minMaxKeys[0], minMaxKeys[1]));
+                        } catch (MetadataNotFoundException me) {
```

Comment:
We should probably have a tool which rewrites current parquet files and add min/max keys. Can we have a issue to track this tool?